tests/unit/lms/services/grouping_test.py

```diff
@@ -29,8 +30,35 @@ def test_upsert_with_parents_inserts(self, svc, db_session):
 
         assert db_session.query(CanvasGroup).one() == test_grouping
 
+    def test_you_cant_upsert_a_grouping_whose_parent_has_a_different_application_instance(
+        self, svc, application_instance_service, db_session
+    ):
+        parent = factories.Course()
+        # GroupingService uses ApplicationInstanceService.get_current() as the
+        # application_instance for the new grouping it inserts.
+        # Here the parent course will have a different application_instance,
+        # which will trigger an IntegrityError when we flush the DB.
+        assert (
+            parent.application_instance
+            != application_instance_service.get_current.return_value
+        )
+
+        svc.upsert_with_parent(
+            tool_consumer_instance_guid="tool_consumer_instance_guid",
+            lms_id="lms_id",
+            lms_name="lms_name",
+            type_=Grouping.Type.CANVAS_GROUP,
+            parent=parent,
+        )
+
+        with pytest.raises(
+            IntegrityError,
+            match='insert or update on table "grouping" violates foreign key constraint "fk__grouping__parent_id__grouping"',
```

Comment:
The `match` here is necessary to ensure that we're catching the `IntegrityError` that we think we're catching. Otherwise something else could be going wrong (there's all sorts of ways to trigger an `IntegrityError`), the code or test could be broken but still passing.